### PR TITLE
Revamp Conditional statement BBEs

### PR DESCRIPTION
**Pull request by DamithaSenevirathne**
## Purpose

This PR will include changes related to,

- if statement BBE
- match statement BBE
- match guard BBE
- Binding patterns in match statement BBE

### CODE REVIEW

File: examples/if-statement/if_statement.bal [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+public function main() {
+    int score = 66;
+    getGrades(score);
+
+    int|string newScore = 77;
+
+    // if statement can be used to type narrowing.
+    if newScore is int {
+        getGrades(newScore);
+    }
```
Comment (hasithaa) [inline examples/if-statement/if_statement.bal:27]: Add else clause and add println to say `score is not an integer`

File: examples/if-statement/if_statement.md [hunk 2]
```diff
@@ -0,0 +1,8 @@
+# If statements
+
+If  statement is used for conditional execution. Conditions following `if` must be boolean and if the 
+condition is true, then the corresponding statement block is executed.
+
```
Comment (hasithaa) [inline examples/if-statement/if_statement.md:5]: We should say about, `else if` and `else` clauses as well. We can't assume users know theses.

File: examples/if-statement/if_statement.md [hunk 3]
```diff
@@ -0,0 +1,8 @@
...
+::: code if_statement.bal :::
+
+::: out if_statement.out :::
```
Comment (hasithaa) [inline examples/if-statement/if_statement.md:8]: Missing related content.

File: examples/booleans/booleans.bal [hunk 4]
```diff
@@ -1,16 +1,9 @@
...
-        io:println(1);
-    } else {
-        io:println(2);
-    }
-}
+    boolean flag = true;
+    io:println(flag);
+
+    // following example will output `false`
+    io:println(3 < 2);
```
Comment (hasithaa) [inline examples/booleans/booleans.bal:8]: Break this into two variable definitions and compare. `3 < 2` Always generates a warning.

File: examples/match-guard-in-match-statements/match_guard_in_match_statements.md [hunk 5]
```diff
@@ -0,0 +1,10 @@
+# Match guard in match statements
+
+A match-guard is an expression that is used in a match clause to determine whether the clause should be executed. A match clause will only be executed if its match-guard evaluates to true
```
Comment (hasithaa) [inline examples/match-guard-in-match-statements/match_guard_in_match_statements.md:3]:
```suggestion
A match-guard is an expression that is used in a match clause to determine whether the clause should be executed. A match clause will only be executed if its match-guard evaluates to true.
```

File: examples/match-guard-in-match-statements/match_guard_in_match_statements.bal [hunk 6]
```diff
@@ -0,0 +1,40 @@
+import ballerina/io;
+
+const switchStatus = "ON";
+
+function matchValue(any val, boolean isObstructed, float powerPercentage) returns string {
+    // The value of the `val` variable is matched against the given value match patterns.
+    match val {
+        // Match guard `if (!isObstructed)` is used
+        1 if (!isObstructed) => {
```
Comment (hasithaa) [inline examples/match-guard-in-match-statements/match_guard_in_match_statements.bal:9]: Remove parenthesis

File: examples/match-guard-in-match-statements/match_guard_in_match_statements.bal [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+            // This block will execute if `!isObstructed` is true
+            return "Move forward";
+        }
+        // use `|` to match more than one value
+        2|3 => {
+            return "Turn";
+        }
+        // Match guard `if (25.0 < powerPercentage)` is used
+        4 if (25.0 < powerPercentage) => {
```
Comment (hasithaa) [inline examples/match-guard-in-match-statements/match_guard_in_match_statements.bal:18]: Remove parenthesis.

**Review by hasithaa — CHANGES_REQUESTED** [review]

**Comment by hasithaa** [comment]
@praneesha will you able to check the PR.

File: examples/if-statement/if_statement.bal [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+}
+
+public function main() {
+    int score = 66;
+    string grade = getGrades(score);
+    io:println(grade);
+
+    int|string newScore = 77;
+
+    // if statement can be used to type narrowing.
```
Comment (lasinicl) [inline examples/if-statement/if_statement.bal:26]:
```suggestion
    // The `if` statement can be used for type narrowing.
```

**Review by hasithaa — APPROVED** [review]
Looks Ok now. Given examples serves the purpose. But I think we can come up with better examples in the next iteration.

File: examples/if-statement/if_statement.md [hunk 9]
```diff
@@ -0,0 +1,13 @@
+# If statements
+
+If  statement is used for conditional execution. Conditions following `if` must be boolean and if the 
+condition is true, then the corresponding statement block is executed. if `if` statement condition is false then 
+one of the following `else if` block executes if particular `else if` statement condition is true. If none of the 
+conditions are true then `else` block executes.
```
Comment (lasinicl) [inline examples/if-statement/if_statement.md:6]: Let's write the description in a single line.

File: examples/match-guard-in-match-statements/match_guard_in_match_statements.bal [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+
+function matchValue(any val, boolean isObstructed, float powerPercentage) returns string {
+    // The value of the `val` variable is matched against the given value match patterns.
+    match val {
+        // Match guard `if !isObstructed` is used.
+        1 if !isObstructed => {
+            // This block will execute if `!isObstructed` is true.
+            return "Move forward";
+        }
+        // use | to match more than one value
```
Comment (lasinicl) [inline examples/match-guard-in-match-statements/match_guard_in_match_statements.bal:13]:
```suggestion
        // Use `|` to match more than one value.
```

File: examples/match-guard-in-match-statements/match_guard_in_match_statements.bal [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+        2|3 => {
+            return "Turn";
+        }
+        // Match guard `if 25.0 < powerPercentage` is used.
+        4 if 25.0 < powerPercentage => {
+            // This block will execute if statement `(25.0 < powerPercentage)`` is true.
```
Comment (lasinicl) [inline examples/match-guard-in-match-statements/match_guard_in_match_statements.bal:19]:
```suggestion
            // This block will execute if `(25.0 < powerPercentage)` is true.
```

File: examples/match-guard-in-match-statements/match_guard_in_match_statements.bal [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+            return  "Increase speed";
+        }
+        "STOP" => {
+            return "STOP";
+        }
+        switchStatus => {
+            return "Switch ON";
+        }
+        // Use _ to match type any.
```
Comment (lasinicl) [inline examples/match-guard-in-match-statements/match_guard_in_match_statements.bal:28]:
```suggestion
        // Use `_` to match type `any`.
```

File: examples/match-guard-in-match-statements/match_guard_in_match_statements.md [hunk 13]
```diff
@@ -0,0 +1,14 @@
+# Match guard in match statements
+
+A match-guard is an expression that is used in a match clause to determine whether the clause should be executed. A match clause will only be executed if its match-guard evaluates to true.
+
+A function call is only allowed with an expression in a match-guard when there is no possibility that it can mutate the value being matched.
+
+
```
Comment (lasinicl) [inline examples/match-guard-in-match-statements/match_guard_in_match_statements.md:7]:
```suggestion
```

File: examples/match-statement/match_statement.bal [hunk 14]
```diff
@@ -1,33 +1,35 @@
...
-            return "number";
+function matchValue(any val) returns string {
+    // The value of the `val` variable is matched against the given value match patterns.
+    match val {
+        1 => {
+            return "Move forward";
         }
-        true => {
-            return "boolean";
+        // use | to match more than one value
```
Comment (lasinicl) [inline examples/match-statement/match_statement.bal:11]:
```suggestion
        // Use `|` to match more than one value.
```

File: examples/match-statement/match_statement.bal [hunk 15]
```diff
@@ -1,33 +1,35 @@
...
         }
-        KEY => {
-            return "constant";
-        }
-        0|1 => {
-            return "or";
+        switchStatus => {
+            return "Switch ON";
         }
+        // use _ to match type any
```
Comment (lasinicl) [inline examples/match-statement/match_statement.bal:21]:
```suggestion
        // Use `_` to match type `any`.
```

File: examples/match-guard-in-match-statements/match_guard_in_match_statements.md [hunk 16]
```diff
@@ -0,0 +1,13 @@
+# Match guard in match statements
```
Comment (praneesha) [inline examples/match-guard-in-match-statements/match_guard_in_match_statements.md:1]:
```suggestion
# Match guard in `match` statements
```

File: examples/match-guard-in-match-statements/match_guard_in_match_statements.metatags [hunk 17]
```diff
@@ -0,0 +1,2 @@
+description: This BBE demonstrates match guards in match statement in Ballerina.
```
Comment (praneesha) [inline examples/match-guard-in-match-statements/match_guard_in_match_statements.metatags:1]:
```suggestion
description: This BBE demonstrates match guards in `match` statements of Ballerina.
```

File: examples/match-guard-in-match-statements/match_guard_in_match_statements.md [hunk 18]
```diff
@@ -0,0 +1,13 @@
...
+
+A match-guard is an expression that is used in a match clause to determine whether the clause should be executed. A match clause will only be executed if its match-guard evaluates to true.
```
Comment (praneesha) [inline examples/match-guard-in-match-statements/match_guard_in_match_statements.md:3]:
```suggestion
A match-guard is an expression that is used in a `match` clause to determine whether the clause should be executed. A `match` clause will only be executed if its match-guard evaluates to true.
```

File: examples/match-statement/match_statement.metatags [hunk 19]
```diff
@@ -1,2 +1,2 @@
 description: This BBE demonstrates match statement in Ballerina.
```
Comment (praneesha) [inline examples/match-statement/match_statement.metatags:1]:
```suggestion
description: This BBE demonstrates the `match` statement in Ballerina.
```

File: examples/if-statement/if_statement.bal [hunk 20]
```diff
@@ -0,0 +1,33 @@
...
+    if 0 < score && score < 55 {
+        return "F";
+    } else if 55 <= score && score < 65 {
+        return "C";
+    } else if 65 <= score && score < 75  {
+        return "B";
+    } else if 75 <= score && score <= 100 {
+        return "A";
+    } else {
+        return "invalid grade";
```
Comment (TharushiJay) [inline examples/if-statement/if_statement.bal:15]:
```suggestion
        return "Invalid grade";
```

File: examples/index.json [hunk 21]
```diff
@@ -110,12 +110,26 @@
...
                 "verifyOutput": true,
                 "isLearnByExample": true
+            },
+            {
+                "name": "Match guard in match statement",
+                "url": "match-guard-in-match-statement",
+                "verifyBuild": true,
+                "verifyOutput": true,
+                "isLearnByExample": true
             }
```
Comment (lasinicl) [inline examples/index.json:133]: Shall we add the details of the `Binding patterns in match statement` BBE here? ATM it is under `Binding Patterns` section. My PR removes that entry from the index.json. Hence we just need to add it here under `Conditional Statements`
```
            {
                "name": "Binding patterns in match statement",
                "url": "binding-patterns-in-match-statement",
                "verifyBuild": true,
                "verifyOutput": true,
                "isLearnByExample": true
            }
```

File: examples/binding-patterns-in-match-statement/binding_patterns_in_match_statement.metatags [hunk 22]
```diff
@@ -1,2 +1,2 @@
 description: This BBE demonstrates binding patterns in a `match` statement.
-keywords: ballerina, ballerina by example, bbe, spread operator, binding pattern
+keywords: ballerina, ballerina by example, bbe, spread operator, binding pattern, match
```
Comment (KavinduZoysa) [inline examples/binding-patterns-in-match-statement/binding_patterns_in_match_statement.metatags:2]: The description should be changed.
Please check https://github.com/ballerina-platform/ballerina-distribution/pull/3910#discussion_r1065326880

**Review by praneesha — APPROVED** [review]

**Review by hasithaa — APPROVED** [review]

**Comment by hasithaa** [comment]
@DamithaSenevirathne There is a conflict, Can you resolve it.

File: examples/binding-patterns-in-match-statement/binding_patterns_in_match_statement.metatags (hunk 22 above)
Comment (DamithaSenevirathne) [reply examples/binding-patterns-in-match-statement/binding_patterns_in_match_statement.metatags:2]: Addressed in e7e755142d12ac2d41edf1411c334d63f3fd0b6d

**Pull request merged**